core/src/main/java/net/sourceforge/jnlp/cache/cache/LeastRecentlyUsedCacheIndexHolder.java

```diff
@@ -0,0 +1,146 @@
+package net.sourceforge.jnlp.cache.cache;
+
+import net.adoptopenjdk.icedteaweb.io.FileUtils;
+import net.adoptopenjdk.icedteaweb.logging.Logger;
+import net.adoptopenjdk.icedteaweb.logging.LoggerFactory;
+import net.sourceforge.jnlp.config.InfrastructureFileDescriptor;
+import net.sourceforge.jnlp.util.PropertiesFile;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Holder of the cached LeastRecentlyUsedCacheIndex data.
+ * The responsibility of this class is to make sure that access to the content of the
+ * LeastRecentlyUsedCacheIndex is synchronized and any changes are written back to the file system.
+ */
+class LeastRecentlyUsedCacheIndexHolder {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LeastRecentlyUsedCacheIndexHolder.class);
+
+    private static ReentrantLock lock = new ReentrantLock();
+
+    private final InfrastructureFileDescriptor recentlyUsed;
+
+    private PropertiesFile cachedIndexPropertiesFile;
+    private List<LeastRecentlyUsedCacheEntry> cachedEntries;
+
+    LeastRecentlyUsedCacheIndexHolder(InfrastructureFileDescriptor recentlyUsed) {
+        this.recentlyUsed = recentlyUsed;
+    }
+
+    /**
+     * Executes the code passed in action in a synchronized fashion to allow exclusive access to the index.
+     */
+    void runSynchronized(Consumer<LeastRecentlyUsedCacheIndex> action) {
+        getSynchronized(idx -> {
+            action.accept(idx);
+            return null;
+        });
+    }
+
+    /**
+     * Executes the function passed in action in a synchronized fashion to allow exclusive access to the index.
+     *
+     * @return the result of the passed function.
+     */
+    <T> T getSynchronized(Function<LeastRecentlyUsedCacheIndex, T> action) {
+        lock.lock();
+        try {
+            final PropertiesFile propertiesFile = lockPropertiesFile();
+            try {
+                final LeastRecentlyUsedCacheIndex index = load(propertiesFile);
+                final T result = action.apply(index);
+                if (index.isDirty()) {
+                    store(propertiesFile);
+                }
+                return result;
+            } finally {
+                unlockPropertiesFile(propertiesFile);
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Lock the properties file to have exclusive access.
+     */
+    private PropertiesFile lockPropertiesFile() {
+        final PropertiesFile cacheProperties = getCacheProperties();
+        cacheProperties.lock();
+        return cacheProperties;
+    }
+
+    /**
+     * Unlock the properties file.
+     */
+    private void unlockPropertiesFile(PropertiesFile propertiesFile) {
+        propertiesFile.unlock();
```

Comment:
do we really need this method?

Reply:
compiler will inline, and the calling code looks much nicer with the method instead of the low level detail